File: Android/app/src/main/java/com/auralis/crisisconnect/ai/CrisisSentinelCloudChatStore.kt
package com.auralis.crisisconnect.ai

import android.content.Context
import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.CollectionReference
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreException
import com.google.firebase.firestore.Query
import java.text.Normalizer
import java.util.Locale
import java.util.UUID
import java.util.concurrent.atomic.AtomicLong
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.shareIn
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withTimeoutOrNull
import org.json.JSONArray
import org.json.JSONObject

/** Helpers for the "cloud:{firestoreDocId}" conversation-id scheme. */
object CrisisSentinelConversationIds {
    const val CLOUD_PREFIX = "cloud:"
    fun isCloud(id: String): Boolean = id.startsWith(CLOUD_PREFIX)
    fun cloudDocId(id: String): String = id.removePrefix(CLOUD_PREFIX)
    fun fromCloudDocId(docId: String): String = CLOUD_PREFIX + docId
}

data class CrisisSentinelTurnSeqs(val userSeq: Long, val assistantSeq: Long)

data class CrisisSentinelCloudChatList(
    val summaries: List<CrisisSentinelConversationSummary> = emptyList(),
    val permissionDenied: Boolean = false
)

data class CrisisSentinelCloudMessages(
    val messages: List<CrisisSentinelChatMessage> = emptyList(),
    /** False once the chat doc is confirmed gone (e.g. deleted from the web dashboard). */
    val chatExists: Boolean = true
)

/**
 * Firestore-backed conversation store shared with the web dashboard's Crisis Sentinel chat.
 * Reads/writes the exact collections and field shapes the web client uses
 * (`agencyPanels/{panelKey}/chats/{chatId}` + `messages` subcollection) so online chats appear
 * on both platforms.
 *
 * App-singleton on purpose: several CrisisSentinelViewModel instances coexist (one per
 * NavBackStackEntry), and all Firestore listening must funnel through the shared
 * [chatSummaries]/[messages] flows so each query has at most one active snapshot listener.
 */
class CrisisSentinelCloudChatStore private constructor(context: Context) {

    private val firestore = FirebaseFirestore.getInstance()
    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val storeScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val lastSeq = AtomicLong(0)

    @Volatile
    private var cachedPanelKey: Pair<String, String?>? = null // uid → derived key (null = none)

    private fun currentUid(): String? =
        FirebaseAuth.getInstance().currentUser?.takeUnless { it.isAnonymous }?.uid

    // ------------------------------------------------------------------ panel key

    /**
     * The chat collection's panel segment — must be byte-identical to the web dashboard's
     * `activeAgencyKey` or chats won't line up. Derivation mirrors dashboard/page.tsx:
     * defaultDashboardPanelId → defaultPanelId → agencySlug → agencyKey (all via
     * [normalizeAgencyDocumentId]: trim + "/"→"-" ONLY) → slugified agency name. Null disables
     * sync (never fall back to a guessed panel).
     */
    suspend fun resolvePanelKey(forceRefresh: Boolean = false): String? {
        val uid = currentUid() ?: return null
        if (!forceRefresh) {
            cachedPanelKey?.let { (cachedUid, key) -> if (cachedUid == uid) return key }
            prefs.getString(panelKeyPrefsKey(uid), null)?.takeIf { it.isNotEmpty() }?.let { key ->
                cachedPanelKey = uid to key
                return key
            }
        }
        val snapshot = runCatching {
            firestore.collection("users").document(uid).get().await()
        }.onFailure { Log.w(TAG, "Could not load users/$uid for panel key", it) }.getOrNull()
            ?: return cachedPanelKey?.takeIf { it.first == uid }?.second
        val key = derivePanelKey(snapshot)
        cachedPanelKey = uid to key
        prefs.edit().putString(panelKeyPrefsKey(uid), key.orEmpty()).apply()
        return key
    }

    fun invalidatePanelKey() {
        val uid = cachedPanelKey?.first
        cachedPanelKey = null
        uid?.let { prefs.edit().remove(panelKeyPrefsKey(it)).apply() }
    }

    private fun panelKeyPrefsKey(uid: String) = "panel_key:$uid"

    private fun derivePanelKey(snapshot: DocumentSnapshot): String? {
        normalizeAgencyDocumentId(snapshot.getString("defaultDashboardPanelId"))?.let { return it }
        normalizeAgencyDocumentId(snapshot.getString("defaultPanelId"))?.let { return it }
        normalizeAgencyDocumentId(snapshot.getString("agencySlug"))?.let { return it }
        normalizeAgencyDocumentId(snapshot.getString("agencyKey"))?.let { return it }
        val agencyName = AGENCY_NAME_FIELDS.firstNotNullOfOrNull { field ->
            snapshot.getString(field)?.trim()?.takeIf { it.isNotEmpty() }
        } ?: return null
        return toAgencyKey(agencyName)
    }

    // ------------------------------------------------------------------ seq

    /**
     * Reserves the user+assistant sequence numbers for a turn up front (web parity:
     * microsecond-scaled epoch, user strictly before assistant even when the reply lands
     * seconds later).
     */
    fun reserveTurnSeqs(): CrisisSentinelTurnSeqs {
        return CrisisSentinelTurnSeqs(userSeq = nextSeq(), assistantSeq = nextSeq())
    }

    private fun nextSeq(): Long {
        while (true) {
            val candidate = System.currentTimeMillis() * 1000
            val previous = lastSeq.get()
            val next = if (candidate > previous) candidate else previous + 1
            if (lastSeq.compareAndSet(previous, next)) return next
        }
    }

    // ------------------------------------------------------------------ writes

    /**
     * Awaits a Firestore write with a cap. Write tasks resolve only on SERVER ack, so an
     * unbounded await would hang forever offline even though the write already landed in the
     * local cache and will sync later (this is what makes Edge replies inside synced chats work
     * in airplane mode). Fast failures (permission-denied while online) still propagate; late
     * failures of a queued write only get logged.
     */
    private suspend fun com.google.android.gms.tasks.Task<*>.awaitAckOrQueue() {
        addOnFailureListener { error ->
            Log.w(TAG, "Deferred cloud chat write failed", error)
        }
        withTimeoutOrNull(WRITE_ACK_TIMEOUT_MS) { await() }
    }

    private suspend fun chatsCollection(): CollectionReference {
        val panelKey = resolvePanelKey() ?: throw FirebaseFirestoreException(
            "No panel key",
            FirebaseFirestoreException.Code.FAILED_PRECONDITION
        )
        return firestore.collection("agencyPanels").document(panelKey).collection("chats")
    }

    /**
     * Creates the chat doc and AWAITS it — the messages subcollection rules `get()` the parent
     * chat's `userId`, so the parent must exist before any message write. Returns the doc id.
     */
    suspend fun createChat(firstUserText: String): String {
        val uid = currentUid() ?: throw FirebaseFirestoreException(
            "Not signed in",
            FirebaseFirestoreException.Code.UNAUTHENTICATED
        )
        val chatRef = chatsCollection().document()
        // Web parity: first 30 chars of the first message, trimmed; literal fallback (data,
        // not UI — rendered on web too, so not localized).
        val title = firstUserText.take(30).trim().ifBlank { "Yeni Sohbet" }
        chatRef.set(
            mapOf(
                "userId" to uid,
                "title" to title,
                "createdAt" to FieldValue.serverTimestamp(),
                "updatedAt" to FieldValue.serverTimestamp(),
                "isTemporary" to false
            )
        ).awaitAckOrQueue()
        return chatRef.id
    }

    suspend fun renameChat(chatDocId: String, title: String) {
        val clean = title.trim().take(60)
        if (clean.isEmpty()) return
        chatsCollection().document(chatDocId)
            .update(mapOf("title" to clean, "updatedAt" to FieldValue.serverTimestamp()))
            .await()
    }

    /** Deletes all message docs in pages (Firestore has no recursive delete), then the chat. */
    suspend fun deleteChat(chatDocId: String) {
        val chatRef = chatsCollection().document(chatDocId)
        while (true) {
            val page = chatRef.collection("messages").limit(100).get().await()
            if (page.isEmpty) break
            val batch = firestore.batch()
            page.documents.forEach { batch.delete(it.reference) }
            batch.commit().await()
            if (page.size() < 100) break
        }
        chatRef.delete().await()
    }

    suspend fun deleteMessages(chatDocId: String, messageIds: List<String>) {
        if (messageIds.isEmpty()) return
        val chatRef = chatsCollection().document(chatDocId)
        val batch = firestore.batch()
        messageIds.forEach { batch.delete(chatRef.collection("messages").document(it)) }
        batch.commit().awaitAckOrQueue()
    }

    suspend fun appendUserMessage(chatDocId: String, text: String, seq: Long) {
        val chatRef = chatsCollection().document(chatDocId)
        val batch = firestore.batch()
        batch.set(
            chatRef.collection("messages").document(UUID.randomUUID().toString()),
            mapOf(
                "sender" to "user",
                "text" to text,
                "timestamp" to FieldValue.serverTimestamp(),
                "seq" to seq,
                // Web parity: always present (some web render paths iterate it unguarded).
                "attachments" to emptyList<Any>()
            )
        )
        batch.update(chatRef, mapOf("updatedAt" to FieldValue.serverTimestamp()))
        batch.commit().awaitAckOrQueue()
    }

    suspend fun appendBotMessage(
        chatDocId: String,
        message: CrisisSentinelChatMessage,
        seq: Long,
        groundingMetadataJson: String? = null
    ) {
        val chatRef = chatsCollection().document(chatDocId)
        // Build the field map imperatively and never put nulls — web docs have ABSENT fields,
        // not null fields (their stripUndefined), and web read paths rely on that.
        val fields = mutableMapOf<String, Any>(
            "sender" to "bot",
            "text" to message.text,
            "timestamp" to FieldValue.serverTimestamp(),
            "seq" to seq,
            "isMarkdown" to true
        )
        message.modelName?.let { fields["modelName"] = it }
        message.cardJson?.let { cardJson ->
            runCatching { jsonObjectToMap(JSONObject(cardJson)) }
                .getOrNull()
                ?.let { fields["card"] = it }
        }
        if (message.mapPoints.isNotEmpty()) {
            fields["mapInfo"] = message.mapPoints.map { point ->
                buildMap<String, Any> {
                    put("lat", point.lat)
                    put("lng", point.lng)
                    put("label", point.label)
                    point.details?.let { put("details", it) }
                    point.type?.let { put("type", it) }
                }
            }
        }
        groundingMetadataJson?.let { json ->
            runCatching { jsonObjectToMap(JSONObject(json)) }
                .getOrNull()
                ?.let { fields["groundingMetadata"] = it }
        }
        val batch = firestore.batch()
        batch.set(chatRef.collection("messages").document(message.id), fields)
        batch.update(chatRef, mapOf("updatedAt" to FieldValue.serverTimestamp()))
        batch.commit().awaitAckOrQueue()
    }

    // ------------------------------------------------------------------ listeners

    /**
     * The signed-in user's cloud chats, shared across all collectors with a single Firestore
     * listener. No orderBy in the query (web parity — avoids a composite index); sorted here.
     */
    val chatSummaries: SharedFlow<CrisisSentinelCloudChatList> = callbackFlow {
        val uid = currentUid()
        val panelKey = resolvePanelKey()
        if (uid == null || panelKey == null) {
            trySend(CrisisSentinelCloudChatList())
            awaitClose { }
            return@callbackFlow
        }
        val registration = firestore.collection("agencyPanels").document(panelKey)
            .collection("chats")
            .whereEqualTo("userId", uid)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    Log.w(TAG, "Cloud chat list listener failed", error)
                    trySend(
                        CrisisSentinelCloudChatList(
                            permissionDenied =
                                error.code == FirebaseFirestoreException.Code.PERMISSION_DENIED
                        )
                    )
                    return@addSnapshotListener
                }
                val summaries = snapshot?.documents.orEmpty()
                    .map(::summaryFromChatDoc)
                    .sortedByDescending { it.updatedAtMillis }
                trySend(CrisisSentinelCloudChatList(summaries = summaries))
            }
        awaitClose { registration.remove() }
    }.shareIn(storeScope, SharingStarted.WhileSubscribed(5_000), replay = 1)

    private val messageFlows = object : LinkedHashMap<String, Flow<CrisisSentinelCloudMessages>>(
        4, 0.75f, true
    ) {
        override fun removeEldestEntry(
            eldest: MutableMap.MutableEntry<String, Flow<CrisisSentinelCloudMessages>>
        ): Boolean = size > 4
    }

    /** Live messages of one cloud chat (+ a chat-doc watch so web-side deletion is detected). */
    fun messages(chatDocId: String): Flow<CrisisSentinelCloudMessages> = synchronized(messageFlows) {
        messageFlows.getOrPut(chatDocId) { createMessagesFlow(chatDocId) }
    }

    private fun createMessagesFlow(chatDocId: String): Flow<CrisisSentinelCloudMessages> =
        callbackFlow {
            val panelKey = resolvePanelKey()
            if (panelKey == null) {
                trySend(CrisisSentinelCloudMessages(chatExists = false))
                awaitClose { }
                return@callbackFlow
            }
            val chatRef = firestore.collection("agencyPanels").document(panelKey)
                .collection("chats").document(chatDocId)
            var latestMessages: List<CrisisSentinelChatMessage> = emptyList()
            var chatExists = true
            val chatRegistration = chatRef.addSnapshotListener { snapshot, _ ->
                // Only trust server-confirmed absence (cache misses report !exists too).
                if (snapshot != null && !snapshot.exists() && !snapshot.metadata.isFromCache) {
                    chatExists = false
                    trySend(CrisisSentinelCloudMessages(latestMessages, chatExists = false))
                }
            }
            val messagesRegistration = chatRef.collection("messages")
                .orderBy("timestamp", Query.Direction.ASCENDING)
                .addSnapshotListener { snapshot, error ->
                    if (error != null) {
                        Log.w(TAG, "Cloud messages listener failed", error)
                        return@addSnapshotListener
                    }
                    latestMessages = snapshot?.documents.orEmpty()
                        .mapNotNull(::messageFromDoc)
                        // seq is authoritative: serverTimestamp is null in cache until commit.
                        .sortedBy { it.first }
                        .map { it.second }
                    trySend(CrisisSentinelCloudMessages(latestMessages, chatExists))
                }
            awaitClose {
                chatRegistration.remove()
                messagesRegistration.remove()
            }
        }.shareIn(storeScope, SharingStarted.WhileSubscribed(5_000), replay = 1)

    // ------------------------------------------------------------------ mapping

    private fun summaryFromChatDoc(doc: DocumentSnapshot): CrisisSentinelConversationSummary {
        return CrisisSentinelConversationSummary(
            id = CrisisSentinelConversationIds.fromCloudDocId(doc.id),
            title = doc.getString("title").orEmpty().ifBlank { "Yeni Sohbet" },
            mode = CrisisSentinelUserMode.FieldTeam,
            // Pending serverTimestamp → "now": the write just happened on this device.
            updatedAtMillis = doc.getTimestamp("updatedAt")?.toDate()?.time
                ?: System.currentTimeMillis(),
            lastMessagePreview = null,
            isCloud = true
        )
    }

    /** Returns orderKey→message; null for malformed docs. */
    private fun messageFromDoc(doc: DocumentSnapshot): Pair<Long, CrisisSentinelChatMessage>? {
        val sender = doc.getString("sender") ?: return null
        val isUser = sender == "user"
        val seq = doc.getLong("seq")
        val timestampMillis = doc.getTimestamp("timestamp")?.toDate()?.time
            ?: seq?.let { it / 1000 }
            ?: System.currentTimeMillis()
        val orderKey = seq ?: (timestampMillis * 1000)
        val modelName = doc.getString("modelName")
        val cardJson = (doc.get("card") as? Map<*, *>)?.let { mapToJsonObject(it).toString() }
        val mapPoints = (doc.get("mapInfo") as? List<*>).orEmpty().mapNotNull { entry ->
            val point = entry as? Map<*, *> ?: return@mapNotNull null
            val lat = (point["lat"] as? Number)?.toDouble() ?: return@mapNotNull null
            val lng = (point["lng"] as? Number)?.toDouble() ?: return@mapNotNull null
            CrisisSentinelMapPoint(
                lat = lat,
                lng = lng,
                label = point["label"] as? String ?: "",
                details = point["details"] as? String,
                type = point["type"] as? String
            )
        }
        return orderKey to CrisisSentinelChatMessage(
            id = doc.id,
            role = if (isUser) CrisisSentinelChatRole.User else CrisisSentinelChatRole.Assistant,
            text = doc.getString("text").orEmpty(),
            timestampMillis = timestampMillis,
            source = when {
                isUser -> null
                modelName == EDGE_MODEL_NAME -> CrisisSentinelResponseSource.LocalModel
                else -> CrisisSentinelResponseSource.OnlineModel
            },
            modelName = modelName,
            cardJson = cardJson,
            mapPoints = mapPoints
        )
    }

    companion object {
        private const val TAG = "CrisisSentinelCloud"
        private const val PREFS_NAME = "crisis_sentinel_cloud"
        private const val WRITE_ACK_TIMEOUT_MS = 4_000L

        /** modelName web shows for on-device replies written into cloud chats (data, not UI). */
        const val EDGE_MODEL_NAME = "Crisis Sentinel Edge"

        private val AGENCY_NAME_FIELDS =
            listOf("agencyName", "agency", "authority", "institution", "organization")

        @Volatile
        private var instance: CrisisSentinelCloudChatStore? = null

        fun getInstance(context: Context): CrisisSentinelCloudChatStore =
            instance ?: synchronized(this) {
                instance ?: CrisisSentinelCloudChatStore(context.applicationContext)
                    .also { instance = it }
            }

        /** Web `normalizeAgencyDocumentId`: trim + "/"→"-" only — NO lowercasing/slugify. */
        internal fun normalizeAgencyDocumentId(raw: String?): String? =
            raw?.trim()?.replace("/", "-")?.takeIf { it.isNotEmpty() }

        /** Web `toAgencyKey`: NFKD → strip combining marks U+0300..U+036F → lowercase → slug. */
        internal fun toAgencyKey(name: String): String? =
            Normalizer.normalize(name, Normalizer.Form.NFKD)
                .replace(Regex("[\\u0300-\\u036F]"), "")
                .lowercase(Locale.ROOT)
                .replace(Regex("[^a-z0-9]+"), "-")
                .trim('-')
                .takeIf { it.isNotEmpty() }

        internal fun jsonObjectToMap(json: JSONObject): Map<String, Any> {
            val result = mutableMapOf<String, Any>()
            json.keys().forEach { key ->
                convertJsonValue(json.get(key))?.let { result[key] = it }
            }
            return result
        }

        private fun jsonArrayToList(array: JSONArray): List<Any> = buildList {
            for (index in 0 until array.length()) {
                convertJsonValue(array.get(index))?.let(::add)
            }
        }

        private fun convertJsonValue(value: Any?): Any? = when (value) {
            null, JSONObject.NULL -> null
            is JSONObject -> jsonObjectToMap(value)
            is JSONArray -> jsonArrayToList(value)
            else -> value
        }

        internal fun mapToJsonObject(map: Map<*, *>): JSONObject {
            val json = JSONObject()
            map.forEach { (key, value) ->
                if (key is String) json.put(key, convertMapValue(value))
            }
            return json
        }

        private fun convertMapValue(value: Any?): Any? = when (value) {
            null -> JSONObject.NULL
            is Map<*, *> -> mapToJsonObject(value)
            is List<*> -> JSONArray().also { array ->
                value.forEach { array.put(convertMapValue(it)) }
            }
            else -> value
        }
    }
}

File: iOS/Crisis Connect/Features/CrisisSentinel/CrisisSentinelCloudChatStore.swift
//
//  CrisisSentinelCloudChatStore.swift
//  Crisis Connect
//
//  S3 — reads/writes the SAME Firestore collections the web dashboard uses
//  (`agencyPanels/{panelKey}/chats` + `messages`) so online (cloud) chats appear identically on
//  web and mobile. iOS port of Android's `CrisisSentinelCloudChatStore`.
//
//  The panel segment MUST be byte-identical to the web dashboard's `activeAgencyKey` (page.tsx),
//  or chats won't line up — see `derivePanelKey` / `toAgencyKey` / `normalizeAgencyDocumentId`.
//

import Foundation
import FirebaseAuth
import FirebaseFirestore

struct CrisisSentinelCloudSummary {
    let docId: String
    let title: String
    let updatedAt: Date
}

struct CrisisSentinelTurnSeqs {
    let userSeq: Int64
    let assistantSeq: Int64
}

actor CrisisSentinelCloudChatStore {
    static let shared = CrisisSentinelCloudChatStore()

    /// modelName web shows for on-device replies written into cloud chats (data, not UI).
    static let edgeModelName = "Crisis Sentinel Edge"
    private static let writeAckTimeout: TimeInterval = 4
    private static let agencyNameFields = ["agencyName", "agency", "authority", "institution", "organization"]

    private let db = Firestore.firestore()
    private let defaults = UserDefaults.standard
    private var cachedPanelKey: (uid: String, key: String?)?
    private var lastSeq: Int64 = 0

    private init() {}

    private func currentUid() -> String? { Auth.auth().currentUser?.uid }

    // MARK: - Panel key (byte-parity with web page.tsx)

    func resolvePanelKey(forceRefresh: Bool = false) async -> String? {
        guard let uid = currentUid() else { return nil }
        if !forceRefresh {
            if let cached = cachedPanelKey, cached.uid == uid { return cached.key }
            if let stored = defaults.string(forKey: Self.panelKeyDefaultsKey(uid)), !stored.isEmpty {
                cachedPanelKey = (uid, stored)
                return stored
            }
        }
        let snapshot = try? await db.collection("users").document(uid).getDocument()
        guard let snapshot, snapshot.exists else {
            return cachedPanelKey.flatMap { $0.uid == uid ? $0.key : nil }
        }
        let key = Self.derivePanelKey(snapshot.data() ?? [:])
        cachedPanelKey = (uid, key)
        defaults.set(key ?? "", forKey: Self.panelKeyDefaultsKey(uid))
        return key
    }

    func invalidatePanelKey() {
        if let uid = cachedPanelKey?.uid {
            defaults.removeObject(forKey: Self.panelKeyDefaultsKey(uid))
        }
        cachedPanelKey = nil
    }

    private static func panelKeyDefaultsKey(_ uid: String) -> String { "crisisSentinel.panelKey.\(uid)" }

    /// Resolution order mirrors the web `activeAgencyKey`: defaultDashboardPanelId → defaultPanelId
    /// → agencySlug → agencyKey (each via normalizeAgencyDocumentId) → slugified agency name.
    static func derivePanelKey(_ data: [String: Any]) -> String? {
        for field in ["defaultDashboardPanelId", "defaultPanelId", "agencySlug", "agencyKey"] {
            if let normalized = normalizeAgencyDocumentId(data[field] as? String) {
                return normalized
            }
        }
        let agencyName = agencyNameFields
            .lazy
            .compactMap { (data[$0] as? String)?.trimmingCharacters(in: .whitespaces) }
            .first { !$0.isEmpty }
        guard let agencyName else { return nil }
        return toAgencyKey(agencyName)
    }

    /// Web `normalizeAgencyDocumentId`: trim + "/"→"-" only — NO lowercasing/slugify.
    static func normalizeAgencyDocumentId(_ raw: String?) -> String? {
        guard let trimmed = raw?.trimmingCharacters(in: .whitespaces), !trimmed.isEmpty else { return nil }
        return trimmed.replacingOccurrences(of: "/", with: "-")
    }

    /// Web `toAgencyKey`: NFKD → strip combining marks U+0300..U+036F → lowercase → slug.
    static func toAgencyKey(_ name: String) -> String? {
        let decomposed = name.decomposedStringWithCompatibilityMapping
        let withoutMarks = decomposed.unicodeScalars.filter { !(0x0300...0x036F).contains($0.value) }
        let lowered = String(String.UnicodeScalarView(withoutMarks)).lowercased()
        var slug = ""
        var lastWasDash = false
        for char in lowered {
            if char.isASCII && (char.isLetter || char.isNumber) {
                slug.append(char)
                lastWasDash = false
            } else if !lastWasDash {
                slug.append("-")
                lastWasDash = true
            }
        }
        while slug.hasPrefix("-") { slug.removeFirst() }
        while slug.hasSuffix("-") { slug.removeLast() }
        // Android returns nil (disable sync) rather than web's "default-agency" guess.
        return slug.isEmpty ? nil : slug
    }

    // MARK: - Seq (microsecond-scaled epoch; user strictly before assistant)

    func reserveTurnSeqs() -> CrisisSentinelTurnSeqs {
        CrisisSentinelTurnSeqs(userSeq: nextSeq(), assistantSeq: nextSeq())
    }

    private func nextSeq() -> Int64 {
        let candidate = Int64(Date().timeIntervalSince1970 * 1000) * 1000
        let next = candidate > lastSeq ? candidate : lastSeq + 1
        lastSeq = next
        return next
    }

    // MARK: - Writes

    private func chatsCollection() async throws -> CollectionReference {
        guard let panelKey = await resolvePanelKey() else {
            throw CrisisSentinelOnlineError(kind: .notConfigured, message: "No panel key")
        }
        return db.collection("agencyPanels").document(panelKey).collection("chats")
    }

    /// Creates the chat doc and AWAITS it — the messages subcollection rules `get()` the parent's
    /// userId, so the parent must exist before any message write. Returns the doc id.
    func createChat(firstUserText: String) async throws -> String {
        guard let uid = currentUid() else {
            throw CrisisSentinelOnlineError(kind: .unauthorized, message: "Not signed in")
        }
        let chatRef = try await chatsCollection().document()
        let title = String(firstUserText.prefix(30)).trimmingCharacters(in: .whitespaces)
        try await chatRef.setData([
            "userId": uid,
            "title": title.isEmpty ? "Yeni Sohbet" : title,
            "createdAt": FieldValue.serverTimestamp(),
            "updatedAt": FieldValue.serverTimestamp(),
            "isTemporary": false,
        ])
        return chatRef.documentID
    }

    func renameChat(docId: String, title: String) async throws {
        let clean = String(title.trimmingCharacters(in: .whitespaces).prefix(60))
        guard !clean.isEmpty else { return }
        try await chatsCollection().document(docId).updateData([
            "title": clean, "updatedAt": FieldValue.serverTimestamp(),
        ])
    }

    func deleteChat(docId: String) async throws {
        let chatRef = try await chatsCollection().document(docId)
        while true {
            let page = try await chatRef.collection("messages").limit(to: 100).getDocuments()
            if page.isEmpty { break }
            let batch = db.batch()
            page.documents.forEach { batch.deleteDocument($0.reference) }
            try await batch.commit()
            if page.count < 100 { break }
        }
        try await chatRef.delete()
    }

    func appendUserMessage(docId: String, text: String, seq: Int64) async {
        guard let chatRef = try? await chatsCollection().document(docId) else { return }
        let batch = db.batch()
        batch.setData([
            "sender": "user",
            "text": text,
            "timestamp": FieldValue.serverTimestamp(),
            "seq": seq,
            "attachments": [Any](),
        ], forDocument: chatRef.collection("messages").document(UUID().uuidString))
        batch.updateData(["updatedAt": FieldValue.serverTimestamp()], forDocument: chatRef)
        await commitDeferred(batch)
    }

    /// Deletes specific message docs from a cloud chat — used when regenerating the last response so
    /// the superseded assistant turn(s) don't linger in Firestore and reappear on the web dashboard
    /// or another device. Mirrors Android's CrisisSentinelCloudChatStore.deleteMessages.
    func deleteMessages(docId: String, messageIds: [String]) async {
        let ids = messageIds.filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
        guard !ids.isEmpty, let chatRef = try? await chatsCollection().document(docId) else { return }
        let batch = db.batch()
        for id in ids {
            batch.deleteDocument(chatRef.collection("messages").document(id))
        }
        await commitDeferred(batch)
    }

    func appendBotMessage(
        docId: String,
        messageId: String,
        text: String,
        modelName: String?,
        cardJson: String?,
        mapPoints: [CrisisSentinelMapPoint],
        groundingMetadataJson: String?,
        seq: Int64
    ) async {
        guard let chatRef = try? await chatsCollection().document(docId) else { return }
        // Never write nulls — web docs have ABSENT fields (their stripUndefined), and web read
        // paths rely on that.
        var fields: [String: Any] = [
            "sender": "bot",
            "text": text,
            "timestamp": FieldValue.serverTimestamp(),
            "seq": seq,
            "isMarkdown": true,
        ]
        if let modelName { fields["modelName"] = modelName }
        if let cardJson, let card = Self.jsonObject(cardJson) { fields["card"] = card }
        if !mapPoints.isEmpty {
            fields["mapInfo"] = mapPoints.map { point -> [String: Any] in
                var m: [String: Any] = ["lat": point.lat, "lng": point.lng, "label": point.label]
                if let details = point.details { m["details"] = details }
                if let type = point.type { m["type"] = type }
                return m
            }
        }
        if let groundingMetadataJson, let grounding = Self.jsonObject(groundingMetadataJson) {
            fields["groundingMetadata"] = grounding
        }
        let batch = db.batch()
        batch.setData(fields, forDocument: chatRef.collection("messages").document(messageId))
        batch.updateData(["updatedAt": FieldValue.serverTimestamp()], forDocument: chatRef)
        await commitDeferred(batch)
    }

    /// Commits with a cap. Write tasks resolve only on SERVER ack, so an unbounded await hangs
    /// forever offline even though the write already landed in the local cache and syncs later.
    private func commitDeferred(_ batch: WriteBatch) async {
        let task = Task { try await batch.commit() }
        _ = try? await withTimeout(Self.writeAckTimeout) { try await task.value }
    }

    // MARK: - Listeners (AsyncStream, one snapshot listener each)

    /// The signed-in user's cloud chats. No orderBy in the query (web parity — avoids a composite
    /// index); sorted by the consumer. Yields nil for the "no panel / not signed in" case.
    func summariesStream() -> AsyncStream<[CrisisSentinelCloudSummary]> {
        AsyncStream { continuation in
            let task = Task {
                guard let uid = currentUid(), let panelKey = await resolvePanelKey() else {
                    continuation.yield([])
                    continuation.finish()
                    return
                }
                let registration = db.collection("agencyPanels").document(panelKey)
                    .collection("chats")
                    .whereField("userId", isEqualTo: uid)
                    .addSnapshotListener { snapshot, _ in
                        let summaries = (snapshot?.documents ?? []).map { doc -> CrisisSentinelCloudSummary in
                            CrisisSentinelCloudSummary(
                                docId: doc.documentID,
                                title: (doc.get("title") as? String)?.nilIfBlank ?? "Yeni Sohbet",
                                updatedAt: (doc.get("updatedAt") as? Timestamp)?.dateValue() ?? Date()
                            )
                        }.sorted { $0.updatedAt > $1.updatedAt }
                        continuation.yield(summaries)
                    }
                continuation.onTermination = { _ in registration.remove() }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    /// Live messages of one cloud chat, ordered. Emits nil `chatExists=false` when the chat was
    /// deleted server-side.
    func messagesStream(docId: String) -> AsyncStream<(messages: [CrisisSentinelChatMessage], chatExists: Bool)> {
        AsyncStream { continuation in
            let task = Task {
                guard let panelKey = await resolvePanelKey() else {
                    continuation.yield(([], false))
                    continuation.finish()
                    return
                }
                let chatRef = db.collection("agencyPanels").document(panelKey)
                    .collection("chats").document(docId)
                var chatExists = true
                let chatReg = chatRef.addSnapshotListener { snapshot, _ in
                    if let snapshot, !snapshot.exists, !snapshot.metadata.isFromCache {
                        chatExists = false
                        continuation.yield(([], false))
                    }
                }
                let msgReg = chatRef.collection("messages")
                    .order(by: "timestamp")
                    .addSnapshotListener { snapshot, _ in
                        let mapped = (snapshot?.documents ?? [])
                            .compactMap { Self.messageFromDoc($0) }
                            .sorted { $0.orderKey < $1.orderKey }
                            .map(\.message)
                        continuation.yield((mapped, chatExists))
                    }
                continuation.onTermination = { _ in chatReg.remove(); msgReg.remove() }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    // MARK: - Mapping

    private static func messageFromDoc(
        _ doc: QueryDocumentSnapshot
    ) -> (orderKey: Int64, message: CrisisSentinelChatMessage)? {
        guard let sender = doc.get("sender") as? String else { return nil }
        let isUser = sender == "user"
        let seq = (doc.get("seq") as? NSNumber)?.int64Value
        let tsMillis = (doc.get("timestamp") as? Timestamp).map { Int64($0.dateValue().timeIntervalSince1970 * 1000) }
            ?? seq.map { $0 / 1000 }
            ?? Int64(Date().timeIntervalSince1970 * 1000)
        let orderKey = seq ?? (tsMillis * 1000)
        let modelName = doc.get("modelName") as? String
        var cardJson: String?
        if let card = doc.get("card") as? [String: Any], let data = try? JSONSerialization.data(withJSONObject: card) {
            cardJson = String(data: data, encoding: .utf8)
        }
        let mapPoints: [CrisisSentinelMapPoint] = (doc.get("mapInfo") as? [[String: Any]] ?? []).compactMap { point in
            guard let lat = (point["lat"] as? NSNumber)?.doubleValue,
                  let lng = (point["lng"] as? NSNumber)?.doubleValue else { return nil }
            return CrisisSentinelMapPoint(
                lat: lat, lng: lng,
                label: point["label"] as? String ?? "",
                details: point["details"] as? String,
                type: point["type"] as? String
            )
        }
        let source: CrisisSentinelResponseSource? = isUser
            ? nil
            : (modelName == edgeModelName ? .localModel : .onlineModel)
        let message = CrisisSentinelChatMessage(
            id: doc.documentID,
            role: isUser ? .user : .assistant,
            text: doc.get("text") as? String ?? "",
            timestamp: Date(timeIntervalSince1970: Double(tsMillis) / 1000),
            source: source,
            confidence: nil,
            modelName: modelName,
            cardJson: cardJson,
            mapPoints: mapPoints
        )
        return (orderKey, message)
    }

    private static func jsonObject(_ raw: String) -> [String: Any]? {
        guard let data = raw.data(using: .utf8) else { return nil }
        return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
    }
}

private extension String {
    var nilIfBlank: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

/// Runs `operation` but gives up after `seconds` (returns nil), without cancelling the underlying
/// Firestore write — it stays queued and syncs when possible.
private func withTimeout<T>(_ seconds: TimeInterval, _ operation: @escaping () async throws -> T) async throws -> T? {
    try await withThrowingTaskGroup(of: T?.self) { group in
        group.addTask { try await operation() }
        group.addTask {
            try? await Task.sleep(nanoseconds: UInt64(seconds * 1_000_000_000))
            return nil
        }
        let result = try await group.next() ?? nil
        group.cancelAll()
        return result
    }
}
